### PR TITLE
[virt-controller]: Refactor renderLaunchManifest function

### PR DESCRIPTION
**Pull request by iholder-redhat**
**What this PR does / why we need it**:
This PR breaks apart some of the logic in `renderLaunchManifest` into smaller private helper functions. [This function](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/services/template.go#L383) is almost 1300 line long and very non-cohesive. This PR reduces it into about 500 lines of code and introduces 5 helper functions to achieve it. Every helper function is cohesive - deals with one task (or renders a different manifest aspect).

This is an initial work that breaks apart some of the logic but not all. Follow-up PR(s) can continue this until we have a clear and clean function.

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/virt-controller/services/template.go
Comment (ashleyschuett) [inline pkg/virt-controller/services/template.go:1176]: is it okay to have this defined twice? can it be changed anywhere? if it gets updated one place but not the other will it break something?

Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go:1176]: TBH I didn't understand what you're referring to. What is defined twice?

**Comment by iholder-redhat** [comment]
@ashleyschuett Just to be clear - I didn't write the code you commented on but only moved it around into a helper function to break `renderLaunchManifest` apart. In any case I've addressed them :smile: 

Thanks! :+1:

**Comment by davidvossel** [comment]
/cc @davidvossel 

just so i remember to review this soon

**Review by davidvossel — COMMENTED** [review]
since you're building helper functions for volumes, i'm curious why some of the volumes are still left in the top level function. for example, these

```
       volumes = append(volumes,
                k8sv1.Volume{
                        Name: "virt-bin-share-dir",
                        VolumeSource: k8sv1.VolumeSource{
                                EmptyDir: &k8sv1.EmptyDirVolumeSource{},
                        },
                },
        )
        volumes = append(volumes, k8sv1.Volume{
                Name: "libvirt-runtime",
                VolumeSource: k8sv1.VolumeSource{
                        EmptyDir: &k8sv1.EmptyDirVolumeSource{},
                },
        })
        volumes = append(volumes, k8sv1.Volume{
                Name: "ephemeral-disks",
                VolumeSource: k8sv1.VolumeSource{
                        EmptyDir: &k8sv1.EmptyDirVolumeSource{},
                },
        })
        volumes = append(volumes, k8sv1.Volume{
                Name: "container-disks",
                VolumeSource: k8sv1.VolumeSource{
                        EmptyDir: &k8sv1.EmptyDirVolumeSource{},
                },
        })
        if !vmi.Spec.Domain.Devices.DisableHotplug {
                volumes = append(volumes, k8sv1.Volume{
                        Name: "hotplug-disks",
                        VolumeSource: k8sv1.VolumeSource{
                                EmptyDir: &k8sv1.EmptyDirVolumeSource{},
                        },
                })
        }

```

**Comment by iholder-redhat** [comment]
> since you're building helper functions for volumes, i'm curious why some of the volumes are still left in the top level function.

Thanks for catching that David!

File: pkg/virt-controller/services/template.go
Comment (acardace) [inline pkg/virt-controller/services/template.go:522]: Can't this whole snippet 
```go
networkToResourceMap, err := getNetworkToResourceMap(t.virtClient, vmi)
	if err != nil {
		return nil, err
	}

	// Register resource requests and limits corresponding to attached multus networks.
	for _, resourceName := range networkToResourceMap {
		if resourceName != "" {
			requestResource(&resources, resourceName)
		}
	}

	err = validatePermittedHostDevices(&vmi.Spec, t.clusterConfig)
	if err != nil {
		return nil, err
	}

	if util.IsGPUVMI(vmi) {
		for _, gpu := range vmi.Spec.Domain.Devices.GPUs {
			requestResource(&resources, gpu.DeviceName)
		}
	}

	if util.IsHostDevVMI(vmi) {
		for _, hostDev := range vmi.Spec.Domain.Devices.HostDevices {
			requestResource(&resources, hostDev.DeviceName)
		}
	}
```
be included in `renderLaunchManifestResourceRequirements` ? We're always dealing with resources, I think it makes sense to group them all under the same function.

**Comment by iholder-redhat** [comment]
Hey @fgimenez @dhiller,

As you can see the same functional tests is failing for all compute lanes ([example](https://prow.ci.kubevirt.io/view/gs/kubevirt-prow/pr-logs/pull/kubevirt_kubevirt/6118/pull-kubevirt-e2e-k8s-1.20-sig-compute/1448638386762944512)):

```
 [sig-compute]Configurations
 tests/vmi_configuration_test.go:66
   VirtualMachineInstance definition
   tests/vmi_configuration_test.go:222
     [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
     tests/vmi_configuration_test.go:1001
       should consume hugepages
       vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
         [Serial][test_id:1672]hugepages-1Gi [It]
         vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:43

         Timed out after 360.001s.
         Timed out waiting for VMI testvmi-clncc to enter [Running] phase(s)
         Expected
             <v1.VirtualMachineInstancePhase>: Pending
         to be an element of
             <[]interface {} | len:1, cap:1>: [["Running"]]

         tests/utils.go:2919
```

The problem is that I cannot reproduce it locally as when I try to run this locally I get this message:
```
sig-compute]Configurations
tests/vmi_configuration_test.go:66
  VirtualMachineInstance definition
  tests/vmi_configuration_test.go:222
    [rfe_id:140][crit:medium][vendor:cnv-qe@redhat.com][level:component]with hugepages
    tests/vmi_configuration_test.go:1001
      should consume hugepages 
      vendor/github.com/onsi/ginkgo/extensions/table/table.go:92
        [Serial][test_id:1672]hugepages-1Gi [It]
        vendor/github.com/onsi/ginkgo/extensions/table/table_entry.go:41

        No node with hugepages hugepages-1Gi capacity

        tests/vmi_configuration_test.go:1077
```

I've talked with @dhiller about this and we figured that there is no clear way (e.g. though environment variable etc') to set the usage of `hugepages-1Gi` in the cluster. I've tried to play with things around to enable it without luck.

Can you advise on how to continue?
Thanks!

**Comment by iholder-redhat** [comment]
Hey @maiqueb! Thanks for your review.

While I agree with your points I think this PR had lost its original purpose which is to split renderLaunchManifest() into multiple helper functions. Instead, this PR had been reviewed multiple times by multiple people always finding better ways of styling the code which I didn't even write..

IMHO we can't turn this PR into "make this code perfect", especially where I didn't write the code and only tried to move it around in order to keep things cleaner so we can further refactor in the future. Future PRs can and should improve this even further, but IMHO this PR is drifting away from its original task of splitting a huge function into multiple helper ones.

**Comment by maiqueb** [comment]
> Hey @maiqueb! Thanks for your review.
> 
> While I agree with your points I think this PR had lost its original purpose which is to split renderLaunchManifest() into multiple helper functions. Instead, this PR had been reviewed multiple times by multiple people always finding better ways of styling the code which I didn't even write..

I think that's part of the purpose of code review: find better ways of styling the code, enabling the reader to have an easier time understanding it. It could be argued that you should incorporate those suggestions into the code. 

> 
> IMHO we can't turn this PR into "make this code perfect", especially where I didn't write the code and only tried to move it around in order to keep things cleaner so we can further refactor in the future. Future PRs can and should improve this even further, but IMHO this PR is drifting away from its original task of splitting a huge function into multiple helper ones.

I agree with small iterative improvements, I've even mentioned that in https://github.com/kubevirt/kubevirt/pull/6118#discussion_r737372323. 

I don't agree with everything you say; for instance, you're introducing a function that returns 6 variables. That's making the code worse imo, since it is error prone - it's easy to put things in the wrong order. I've said so, and I think it should be changed. Same with adding a couple of consts that - when correctly named - will indicate to the reader what they are , and how they're supposed to be used. 

You can still split this "huge" function into smaller ones and not have functions that return 6 variables.

I haven't blocked this PR though. I just voiced my concerns that the intermediate state you're proposing has parts that are confusing and error prone; if other reviewers agree with me, I hope you take these into account.

**Comment by iholder-redhat** [comment]
> > Hey @maiqueb! Thanks for your review.
> > While I agree with your points I think this PR had lost its original purpose which is to split renderLaunchManifest() into multiple helper functions. Instead, this PR had been reviewed multiple times by multiple people always finding better ways of styling the code which I didn't even write..
> 
> I think that's part of the purpose of code review: find better ways of styling the code, enabling the reader to have an easier time understanding it. It could be argued that you should incorporate those suggestions into the code.

I agree, except this PR is meant to complete a small refactoring task, not to beautify every part of the code that is moving around

> > IMHO we can't turn this PR into "make this code perfect", especially where I didn't write the code and only tried to move it around in order to keep things cleaner so we can further refactor in the future. Future PRs can and should improve this even further, but IMHO this PR is drifting away from its original task of splitting a huge function into multiple helper ones.
> 
> I agree with small iterative improvements, I've even mentioned that in [#6118 (comment)](https://github.com/kubevirt/kubevirt/pull/6118#discussion_r737372323).

Great

> I don't agree with everything you say; for instance, you're introducing a function that returns 6 variables. That's making the code worse imo, since it is error prone - it's easy to put things in the wrong order. I've said so, and I think it should be changed. Same with adding a couple of consts that - when correctly named - will indicate to the reader what they are , and how they're supposed to be used.
> 
> You can still split this "huge" function into smaller ones and not have functions that return 6 variables.

Let me disagree. I think that a function of almost 2K lines is probably the hardest code to read and understand, and therefore maintain. The first step IMO is to break the logic apart. This is the small incremental change this PR tries to achieve. In a future work we should take your advice and continue beautifying the rest, but I certainly do not agree that a messy 2K lines function is better than a small function which in one of its lines returns 6 values.

> I haven't blocked this PR though. I just voiced my concerns that the intermediate state you're proposing has parts that are confusing and error prone; if other reviewers agree with me, I hope you take these into account.

And thank you for that! All of your points are important and should be addressed in future PRs.

**Comment by maiqueb** [comment]
> > > Hey @maiqueb! Thanks for your review.
> > > While I agree with your points I think this PR had lost its original purpose which is to split renderLaunchManifest() into multiple helper functions. Instead, this PR had been reviewed multiple times by multiple people always finding better ways of styling the code which I didn't even write..
> > 
> > 
> > I think that's part of the purpose of code review: find better ways of styling the code, enabling the reader to have an easier time understanding it. It could be argued that you should incorporate those suggestions into the code.
> 
> I agree, except this PR is meant to complete a small refactoring task, not to beautify every part of the code that is moving around

I don't think my suggestions count as "beautify". It's functional: it helps the reader understand it better. That's important, since code is read a lot more times than it is written.

> 
> > > IMHO we can't turn this PR into "make this code perfect", especially where I didn't write the code and only tried to move it around in order to keep things cleaner so we can further refactor in the future. Future PRs can and should improve this even further, but IMHO this PR is drifting away from its original task of splitting a huge function into multiple helper ones.
> > 
> > 
> > I agree with small iterative improvements, I've even mentioned that in [#6118 (comment)](https://github.com/kubevirt/kubevirt/pull/6118#discussion_r737372323).
> 
> Great
> 
> > I don't agree with everything you say; for instance, you're introducing a function that returns 6 variables. That's making the code worse imo, since it is error prone - it's easy to put things in the wrong order. I've said so, and I think it should be changed. Same with adding a couple of consts that - when correctly named - will indicate to the reader what they are , and how they're supposed to be used.
> > You can still split this "huge" function into smaller ones and not have functions that return 6 variables.
> 
> Let me disagree. I think that a function of almost 2K lines is probably the hardest code to read and understand, and therefore maintain. The first step IMO is to break the logic apart. This is the small incremental change this PR tries to achieve. In a future work we should take your advice and continue beautifying the rest, but I certainly do not agree that a messy 2K lines function is better than a small function which in one of its lines returns 6 values.

Let me clarify: I'm not saying a "messy 2K lines function is better than a small function which in one of its lines returns 6 values". I'm saying that one of the functions that you're proposing has a signature that is error prone, and hard to follow. 

As such, I've asked to consider updating it. I also don't buy the argument that "a small function which in one of its lines returns 6 values" ... That single line you mention ends up being the one that specifies the contract between the implementation and the caller of the function. 

> 
> > I haven't blocked this PR though. I just voiced my concerns that the intermediate state you're proposing has parts that are confusing and error prone; if other reviewers agree with me, I hope you take these into account.
> 
> And thank you for that! All of your points are important and should be addressed in future PRs.

File: pkg/virt-controller/services/template.go
Comment (maiqueb) [inline pkg/virt-controller/services/template.go]: is this supposed to be in this commit ? where did these come from ?

File: pkg/virt-controller/services/template.go
Comment (maiqueb) [inline pkg/virt-controller/services/template.go]: please squash this into the correct commit, https://github.com/kubevirt/kubevirt/pull/6118/commits/8bbaad210ed5c3aee0030bcb89e61783cace6d02

File: pkg/virt-controller/services/template.go
Comment (maiqueb) [reply pkg/virt-controller/services/template.go:565]: This comment was not acted upon, btw - i.e. why is this named `extra...` ?

File: pkg/virt-controller/services/template.go
Comment (maiqueb) [inline pkg/virt-controller/services/template.go]: This code was introduced in https://github.com/kubevirt/kubevirt/pull/6118/commits/3ff727b552e3c6d86aceba6eddf37f1765439c90. 

Please squash this change there.

File: pkg/virt-controller/services/template.go
Comment (maiqueb) [inline pkg/virt-controller/services/template.go:155]: I wonder if the `bool` is actually needed - you could, iiuc, just use `map[string]struct{}` instead, since you never set `false` into the map.

No biggie, though, I won't insist.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Huge +1 on the last commit of the stack.

I'm just asking for a couple of commits here to be squashed into others, since they're fixes to errors introduced within this very same PR.

Once that's taken care of, you have my lgtm.

Thanks.

File: pkg/virt-controller/services/template.go
Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go]: done

File: pkg/virt-controller/services/template.go
Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go:565]: Although your comment is great it is not about a code I wrote but rather on a code that I moved from place to place, therefore this is outside of the scope of this PR and better be addressed in future changes.

File: pkg/virt-controller/services/template.go
Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go]: done

File: pkg/virt-controller/services/template.go
Comment (iholder-redhat) [reply pkg/virt-controller/services/template.go:155]: Same as above:
although your comment is great it is not about a code I wrote but rather on a code that I moved from place to place, therefore this is outside of the scope of this PR and better be addressed in future changes.

**Comment by iholder-redhat** [comment]
> Huge +1 on the last commit of the stack.
> 
> I'm just asking for a couple of commits here to be squashed into others, since they're fixes to errors introduced within this very same PR.
> 
> Once that's taken care of, you have my lgtm.
> 
> Thanks.

Thanks, addressed your feedback :+1:

File: pkg/virt-controller/services/template.go
Comment (maiqueb) [reply pkg/virt-controller/services/template.go:155]: Perfectly fine.

**Review by maiqueb — APPROVED** [review]
Thanks for your patience.

**Comment by iholder-redhat** [comment]
> Thanks for your patience.

And thank you very much for your review and great ideas!
I sincerely think your ideas are very important and should be addressed in the future

**Review by enp0s3 — COMMENTED** [review]
Thank you!
/approve

**Comment by rmohr** [comment]
/hold

refactors in that area are critical. I think @davidvossel wanted to have a look.

**Comment by iholder-redhat** [comment]
> /hold
> 
> refactors in that area are critical. I think @davidvossel wanted to have a look.

**Comment by iholder-redhat** [comment]
> /hold
> 
> refactors in that area are critical. I think @davidvossel wanted to have a look.

Of course.
@davidvossel In two words this PR splits the function to several helper functions that are more cohesive and focused on a logical rendering sub-task, i.e. no logical changes are expected from this PR, only pure refactoring.

Please let me know what you think of this.

**Review by davidvossel — COMMENTED** [review]
While this does break up logic into smaller functions and reduce overall size of the renderLaunchManifest function, I don't think the new structure necessarily makes this function any easier to work in or understand. 

Here is what I mean.

Let's take for example volumes. We still have volume information being generated and spread out all throughout this function and nested functions. Except now instead of having a single set of slices to follow at the top of the renderLaunchManifest function, we have multiple objects and object pointers being passed around  through layers of nested functions. 

The rendering of volumes isn't concise either. For example, we're applying fields from the `volumeInfo` object to container specis and then modifying the volumeInfo object more later on.

**Comment by iholder-redhat** [comment]
> While this does break up logic into smaller functions and reduce overall size of the renderLaunchManifest function, I don't think the new structure necessarily makes this function any easier to work in or understand.
> 
> Here is what I mean.
> 
> Let's take for example volumes. We still have volume information being generated and spread out all throughout this function and nested functions. Except now instead of having a single set of slices to follow at the top of the renderLaunchManifest function, we have multiple objects and object pointers being passed around through layers of nested functions.
> 
> The rendering of volumes isn't concise either. For example, we're applying fields from the `volumeInfo` object to container specs and then modifying the volumeInfo object more later on.

Thanks @davidvossel.

IMHO having a HUGE function that does a lot of unrelated things is the worst code to maintain. I understand that further refactoring is needed here and that the division to sub-function is not perfectly cohesive, but I do think that it makes the situation much better. At least renderLaunchManifest's code is readable and it's pretty much clear where everything is being done.

This is meant to be a first step towards refactoring this function. I think we should perform incremental changes and the first obvious step IMO is to split this function into more or less cohesive helper functions.

Another note: I honestly think it's problematic that as a project such PRs are so difficult to push forward. IMO small incremental PRs to refactor the code should happen on a regular basis. Instead, this PR is out there for months now for what I thought of as a tiny incremental change that will also encourage others to further refactor it.

How do you think we should continue?

**Comment by iholder-redhat** [comment]
> I think your direction is a great start. What I'd like to see is that the top level render function only uses volume information rendered by helper functions. So move everything involved with rendering volumes out of the top level function.
> 
> It's really tough to untangle all this

**Comment by iholder-redhat** [comment]
Thanks @davidvossel.

> I think my primary issue with the refactor is it doesn't completely refactor the rendering of the volumes, instead we're split between some volumes being rendered in the giant top level function and the majority being rendered in helper functions.

So, in other words - this refactoring work is not yet completed and not perfect, we have more work do to. I Agree.

> I think your direction is a great start.

So why don't we merge it?

> What I'd like to see is that the top level render function only uses volume information rendered by helper functions. So move everything involved with rendering volumes out of the top level function.
>
> It's really tough to untangle all this

If it's really tough, why insist on doing all the work in the same PR?

Usually, small incremental changes are encouraged (rightfully), I honestly don't understand why we take the complete opposite approach when it comes to refactoring. You said this PR is a "great start" - so I assume you agree that it makes things better, not worse - so IMHO we should merge it and then think on how to further improve this.

And again, I think about it from the perspective of how do we want to behave as a community. Let's be honest - refactoring work is not very fun to do, and usually it's not an impressive task for your resume. It's a lot easier to see a 2K lines function and just add to it the stuff you need for that feature you're working on. How do we encourage people to stop what they're doing and say "I'm gonna spend some time on making things better, even it's outside of my task's scope"? If refactoring means to start a very heavy task which will takes months to complete and the expectation is that you do it perfectly or it won't get merged - IMHO it means that de-facto no one will do such work.As a proof we can see that one of our most important functions that are in the heart of are flow hadn't been refactored and just keeps growing and growing...

The main goal of this PR is not to perfectly refactor everything - but to be an **initial work** that will encourage others and myself to continue the refactoring effort, as PR description says.

**Comment by alicefr** [comment]
I admit I missed completely this PR, so it was great that it was mentioned during the community meeting. But I'd like to give some good reasons why a refactoring of this function could be very beneficial. We have a couple of situations like WaitForFirstConsumer or the attachment pod for hotplug where we need to create a pod with the same scheduling constraints as the VM. I'm also working on a similar feature where I need this. If we can separate the part of the renderLaunchManifest that influences the scheduling of the VM, those functions could be reused for the creation of these helper pods.

**Comment by rmohr** [comment]
> The main goal of this PR is not to perfectly refactor everything - but to be an **initial work** that will encourage others and myself to continue the refactoring effort, as PR description says.

Taking into account what David said in https://github.com/kubevirt/kubevirt/pull/6118#issuecomment-970721810, I would suggest to take a small domain (like e.g. volume rendering) and try to factor it out. This will likely also involve slight restructuring of other parts of the function and will lead to the next step, by try to keep all other changes at a minimum.

This would still be a (focused) "initial work" which can still encourage others. If you want to do big refactoring at once (also fine in principal), a clear idea and vision with clear benefits of the restructuring is required. If a refactoring does for instance not clearly improve e.g. isolation, and instead introduce more paths to track, then it is probably not worth merging it. Especially when it is a big refactor.

**Comment by brybacki** [comment]
> I admit I missed completely this PR, so it was great that it was mentioned during the community meeting. But I'd like to give some good reasons why a refactoring of this function could be very beneficial. We have a couple of situations like WaitForFirstConsumer or the attachment pod for hotplug where we need to create a pod with the same scheduling constraints as the VM. I'm also working on a similar feature where I need this. If we can separate the part of the renderLaunchManifest that influences the scheduling of the VM, those functions could be reused for the creation of these helper pods.

That is right, and you can clearly see that WaitForFirstConsumer was added as a boolean (I did it) flag because there was no easy way to reuse/extend the function. So refactoring will help with tasks like that, and with testing new additions.

**Comment by alicefr** [comment]
> > I admit I missed completely this PR, so it was great that it was mentioned during the community meeting. But I'd like to give some good reasons why a refactoring of this function could be very beneficial. We have a couple of situations like WaitForFirstConsumer or the attachment pod for hotplug where we need to create a pod with the same scheduling constraints as the VM. I'm also working on a similar feature where I need this. If we can separate the part of the renderLaunchManifest that influences the scheduling of the VM, those functions could be reused for the creation of these helper pods.
> 
> That is right, and you can clearly see that WaitForFirstConsumer was added as a boolean (I did it) flag because there was no easy way to reuse/extend the function. So refactoring will help with tasks like that, and with testing new additions.

At least, isolating the volumes (volumeMounts and deviceMounts), node selector and the resources assigned to the pod (like huge pages and host devices) can be very useful.

**Pull request closed without merging**
